File: go/quick_sort/quick_sort.go
package main

import "fmt"

func partition(arr []int, lo int, hi int) int {
	pivot := arr[hi]
	idx := lo - 1

	for i := lo; i < hi; i++ {
		if arr[i] <= pivot {
			idx++
			tmp := arr[i]
			arr[i] = arr[idx]
			arr[idx] = tmp
		}
	}

	idx++

	arr[hi] = arr[idx]
	arr[idx] = pivot

	return idx
}

func quickSort(arr []int, lo int, hi int) {
	if lo >= hi {
		return
	}

	pivotIdx := partition(arr, lo, hi)

	quickSort(arr, lo, pivotIdx-1)
	quickSort(arr, pivotIdx+1, hi)
}

func main() {
	arrTest := []int{8, 6, 2, 3, 9}
	quickSort(arrTest, 0, len(arrTest)-1)
	fmt.Println(arrTest)
}

File: c_c++/quickSort/quickSort.cpp
#include <iostream>
using namespace std;
 
int partition(int arr[], int lo, int hi)
{
    int pivot = arr[hi];
    int idx = lo - 1;

    for (int i = lo; i < hi; i++) {
        if (arr[i] <= pivot) {
            ++idx;
            int tmp = arr[i];
            arr[i] = arr[idx];
            arr[idx] = tmp;
        }
    }
 
   ++idx;
   arr[hi] = arr[idx];
   arr[idx] = pivot;
 
    return idx;
}
 
void quickSort(int arr[], int lo, int hi)
{
 
    // base case
    if (lo >= hi)
        return;
 
    // partitioning the array
    int p = partition(arr, lo, hi);
 
    // Sorting the left part
    quickSort(arr, lo, p - 1);
 
    // Sorting the right part
    quickSort(arr, p + 1, hi);
}
 
int main()
{
 
    int arr[] = { 9, 3, 4, 2, 1, 8 };
    int n = 6;
 
    quickSort(arr, 0, n - 1);
 
    for (int i = 0; i < n; i++) {
        cout << arr[i] << " ";
    }
 
    return 0;
}